### PR TITLE
Added path flag and respective logic

### LINKED ISSUE #1587: 'odo list' command should provide option to show not deployed components in local filesystem along with deployed ones
[kind/Feature]

## Which functionality do you think we should add?
In odo up to v0.0.20 new component is always deployed to the cluster. That means after successful 'odo create'  command 'odo list' would always list created component. With new workflow that would not work that way anymore, because new component created only in local file-system and not deployed to the cluster.

Would be good to support that in new odo without change, by saving new not deployed component's location in ~/.odo configuration and then inlude them in `odo list` or at least provide search for component in the specified local file system folder(s) with recursion support.

## Why is this needed?
To allow developers and tools see list of all locally created not deployed yet components.

## Workaround 
Devs or tools would have to iterate over local file-system folder and use `odo describe --context folder/path` to test if there component in specified `folder/path`.

### PR DESCRIPTION
**Pull request by surajnarwade**
## Was the change discussed in an issue?
fixes #1587

## How to test changes?

```
[snarwade@redsuraj github.com]$ odo list --context openshift/nodejs-ex --path openshift
NAME            TYPE       SOURCE     STATE            CONTEXT
php-esgq        php        ./         not deployed     /home/snarwade/go/src/github.com/openshift/cakephp-ex
httpd-nesr      httpd      ./         not deployed     /home/snarwade/go/src/github.com/openshift/httpd-ex
nodejs-rtsc     nodejs     ./         deployed         /home/snarwade/go/src/github.com/openshift/nodejs-ex
ruby-kept       ruby       ./         deployed         /home/snarwade/go/src/github.com/openshift/ruby-ex
```

### CODE REVIEW

**Comment by surajnarwade** [comment]
@kadel , @dgolovin please try this and check output and provide feedback (though this WIP PR)
(I havn't looked at sanity checks yet :see_no_evil: )

**Comment by cdrage** [comment]
This is **awesome** and a welcomed change, however, could we align it so that `odo list` shows similarly to `odo list --path` ?

```sh
▶ odo list
ACTIVE     NAME            TYPE
*          nodejs-trym     nodejs

~/nodejs-ex  master ✗                                                                                                                                                                                                                                                                                                                                               223d ◒  
▶ odo list --path ~/nodejs-ex
NAME            TYPE       SOURCE     STATE        CONTEXT
nodejs-trym     nodejs     ./         deployed     /home/wikus/nodejs-ex
```

IMO, I think that having State as well as Context in the normal `odo list` would be beneficial.

What about the `ACTIVE` column too?

Side-note: Deployed I think should be capitalized.

**Comment by surajnarwade** [comment]
@cdrage thanks for the feedback, I added `STATE` for normal `odo list` which will be always `Deployed`, I am confused how to get `CONTEXT` field for normal `odo list`
cc @kadel

**Comment by cdrage** [comment]
@surajnarwade Looks awesome:

```sh
▶ odo list
ACTIVE     NAME                    TYPE       SOURCE                                     STATE
           example-dotnet-sjco     dotnet     file:///tmp/binary/example.war             Deployed
           example-java-fiov       java       file://example.war                         Deployed
           foobar                  nodejs     file://./                                  Deployed
           java-fwwm               java       file://./                                  Deployed
           nodejs-nuvb             nodejs     file://./                                  Deployed
*          nodejs-ynrj             nodejs     https://github.com/openshift/nodejs-ex     Deployed
```

As per https://en.wikipedia.org/wiki/File_URI_scheme the file:// scheme is correct.

I don't see anything wrong with this and I like the implementation.

Another issue I found.

If you create a component and don't deploy it, it doesn't show up as "Undeployed" or "Not deployed"..

Ex:

```sh
odo create nodejs --git https://github.com/openshift/nodejs-ex
odo list
```

**Comment by surajnarwade** [comment]
@cdrage, as of now, I worked on only `--path` functionality as you can see.

`odo list`  (without --path) gathers component info only by talking to cluster, so it not aware of `not Deployed` components from config file which it is using

**Comment by cdrage** [comment]
@surajnarwade Ah, we should then remove the deployed column then if it's not detecting it, or can you add the functionality?

**Comment by mik-dass** [comment]
I think we don't need this active column anymore 

```
[mrinaldas@localhost occlient]$ odo list --path ~/ --context ~/nodejs-ex/
ACTIVE     NAME            TYPE       SOURCE                                  STATE            CONTEXT
           nodejs-knib     nodejs     https://github.com/sclorg/nodejs-ex     not Deployed     /home/mrinaldas/nodejs-3.5
*          nodejs-eakr     nodejs     ./                                      not Deployed     /home/mrinaldas/nodejs-ex
```

File: pkg/odo/cli/component/list.go [hunk 1]
```diff
@@ -52,6 +57,65 @@ func (lo *ListOptions) Validate() (err error) {
 
 // Run has the logic to perform the required actions as part of command
 func (lo *ListOptions) Run() (err error) {
+	var ap []component.Component
+	if lo.pathFlag != "" {
+		err := filepath.Walk(lo.pathFlag, func(path string, f os.FileInfo, err error) error {
```
Comment (kadel) [inline pkg/odo/cli/component/list.go]: Can we put this in some separate function? This `if` statement is getting huge.

Comment (surajnarwade) [reply pkg/odo/cli/component/list.go]: sure

**Comment by surajnarwade** [comment]
@kadel , WIP :smile:

**Comment by kadel** [comment]
> @kadel , WIP 😄

NP, just making sure that it doesn't get lost ;-)

**Comment by surajnarwade** [comment]
@kadel , that will be possible but it will take a time to crawl, probably will not be the good UX, we can add if user asks for it ? WDYT ?

**Comment by kadel** [comment]
> @kadel , that will be possible but it will take a time to crawl, probably will not be the good UX, we can add if user asks for it ? WDYT ?

The user that we are writing this feature for already asked for it ;-) https://github.com/openshift/odo/issues/1587#issuecomment-497820283

**Comment by dgolovin** [comment]
@kadel json without project and application should work, because first we build cluster/project/app hierarchy. Then request list of components for each application using `odo list --project projectName --app appName`, that means we know project and app names for listed components.

**Comment by surajnarwade** [comment]
@dgolovin , can you please test this PR and check whether it is working as per your expectation ?

File: pkg/component/types.go [hunk 2]
```diff
@@ -31,7 +31,8 @@ type ComponentList struct {
 
 // ComponentStatus is Status of components
 type ComponentStatus struct {
-	Active           bool                `json:"active"`
+	Context          string              `json:"context,omitempty"`
+	State            bool                `json:"state"`
```
[the remark this answers is not in the document]
Comment (surajnarwade) [reply pkg/component/types.go]: ok :+1:

File: pkg/component/component.go [hunk 3]
```diff
@@ -900,6 +901,31 @@ func List(client *occlient.Client, applicationName string) (ComponentList, error
 	return compoList, nil
 }
 
+// ListIfPathGiven lists all available component in given path directory
+func ListIfPathGiven(path string, client *occlient.Client) (ComponentList, error) {
```
[the remark this answers is not in the document]
Comment (surajnarwade) [reply pkg/component/component.go]: :+1:  for client place,
/me bad at naming, any suggestions for the name ?

**Comment by dgolovin** [comment]
I have these issues:
1. `odo list --path` always required to be running from existing component's context folder. What if I don't know what folder has a component and just want to search components for specific location.
2. `odo list --path ./not/existing/path` fails with stacktrace 
image
3. `odo list --path location1:location2:location3` fails with error similar to shown above

**Comment by surajnarwade** [comment]
> odo list --path always required to be running from existing component's context folder. What if I don't know what folder has a component and just want to search components for specific location.

@dgolovin this will resolved by upcoming @mik-dass  PR

/me looking into issue 2 and 3.

**Comment by surajnarwade** [comment]
> odo list --path ./not/existing/path fails with stacktrace

I am not able to reproduce this, can you please provide context also

> odo list --path location1:location2:location3 fails with error similar to shown above

for this there are possible solutions:

1. `odo list --path a --path b`
2. `odo list --path a:b`
3. `odo list --path a,b`

what do you prefer ? @dgolovin , @kadel

**Comment by kadel** [comment]
> > odo list --path ./not/existing/path fails with stacktrace
> 
> I am not able to reproduce this, can you please provide context also

@surajnarwade maybe it is windows specific?

> 
> > odo list --path location1:location2:location3 fails with error similar to shown above
> 
> for this there are possible solutions:
> 
> 1. `odo list --path a --path b`
> 2. `odo list --path a:b`
> 3. `odo list --path a,b`
> 
> what do you prefer ? @dgolovin , @kadel

1. this is how cobra does it when you use `StringSliceVar` or `StringSliceVarP`.

**Comment by surajnarwade** [comment]
> this is how cobra does it when you use StringSliceVar or StringSliceVarP.

right, I am following the same, will send soon

**Comment by surajnarwade** [comment]
@dgolovin , @kadel , I updated the PR with `--path a --path b` feature, can you please check and test it ?

**Comment by kadel** [comment]
> > odo list --path ./not/existing/path fails with stacktrace
> 
> I am not able to reproduce this, can you please provide context also

@surajnarwade From the stacktrace it looks like `f` can  be `nil`  for some files (in `filepath.Walk(path, func(path string, f os.FileInfo, err error)` )

**Comment by surajnarwade** [comment]
@kadel , yeah
/me adding case for the same

**Comment by dgolovin** [comment]
@kadel @surajnarwade Using `--path /path1 --path /path2 --path ...`  would require a extra typing, but should work.

I would go with `--path path1:path2:path3` option because it is less typing. It is also 7 characters less per folder and would let handle more folders in one command considering limited command length ~8Kb on latest Windows.
This option seems to be very easy to implement with [SplitList](https://golang.org/pkg/path/filepath/#example_SplitList) function, which would support Windows/Unix file lists transparently.

For extreme cases feeding file list through a file would be required, but that should not be needed soon ;)

**Comment by surajnarwade** [comment]
@dgolovin ack

**Comment by mik-dass** [comment]
It's not working if two path are mentioned and in both `~` is used for representing the home folder.

```
[mrinaldas@localhost ~]$ odo list --context /home/mrinaldas/nodejs-3.5/ --path ~/nodejs-ex/:~/prac/java
NAME            TYPE       SOURCE     STATE          CONTEXT
nodejs-rpqm     nodejs     ./         Not Pushed     /home/mrinaldas/nodejs-ex
```
In the above example, I have a odo component folder in `~/prac/java`

By further digging into the code, I discovered that the `~` for the second argument is not resolved to the home directory.

```
[mrinaldas@localhost ~]$ odo list --context /home/mrinaldas/nodejs-3.5/ --path ~/nodejs-ex/:~/prac/java
/home/mrinaldas/nodejs-ex
~/prac/java
NAME            TYPE       SOURCE     STATE          CONTEXT
nodejs-rpqm     nodejs     ./         Not Pushed     /home/mrinaldas/nodejs-ex
```

**Comment by kadel** [comment]
> @kadel @surajnarwade Using `--path /path1 --path /path2 --path ...` would require a extra typing, but should work.
> 
> I would go with `--path path1:path2:path3` option because it is less typing. It is also 7 characters less per folder and would let handle more folders in one command considering limited command length ~8Kb on latest Windows.
> This option seems to be very easy to implement with [SplitList](https://golang.org/pkg/path/filepath/#example_SplitList) function, which would support Windows/Unix file lists transparently.
> 
> For extreme cases feeding file list through a file would be required, but that should not be needed soon ;)

How many paths do you plan to provide? It doesn't look right to me if you need to provide more than 2 or 3.

**Comment by surajnarwade** [comment]
@mik-dass , I found the problem, go doesn't handle `~`  properly, so we can use `https://github.com/mitchellh/go-homedir` to expand the path if `~` is given

**Comment by kadel** [comment]
I've tried this PR with https://github.com/openshift/odo/pull/1792, and I  still can't run this outside of the odo component directory :-/

```
▶ odo component list  --path .
 ✗  the current directory does not represent an odo component.
Maybe use 'odo create' to create component here or switch to directory with a component
```

@surajnarwade this is slightly different than what is described in https://github.com/openshift/odo/issues/1646. You need to make sure `--path` doesn't require a component directory.

**Comment by dgolovin** [comment]
> How many paths do you plan to provide? It doesn't look right to me if you need to provide more than 2 or 3.

@akadel I would provide as many as added to vscode workspace for now because I really like the opportunity to do it in one odo call. We can also do iteration with one call per added workspace folder. Or even add a check for command length, if it is 
over the max we would bundle as much as we can in one command and do it in several calls.

**Comment by surajnarwade** [comment]
@kadel , https://github.com/openshift/odo/pull/1792 will handle running component list command without requiring component directory

**Comment by kadel** [comment]
> @kadel , #1792 will handle running component list command without requiring component directory

Not in a way this PR requires.

File: pkg/odo/genericclioptions/context.go [hunk 4]
```diff
@@ -118,9 +118,11 @@ func getValidConfig(command *cobra.Command) (*config.LocalConfigInfo, error) {
 		// Gather nessasary info
 		p := command.Parent()
 		r := command.Root()
+		fmt.Println("root is", r.Name(), "parent is", p.Name())
```
Comment (kadel) [inline pkg/odo/genericclioptions/context.go]: I suspect that this is debugging output left accidentally

File: pkg/odo/genericclioptions/context.go [hunk 5]
```diff
@@ -118,9 +118,11 @@ func getValidConfig(command *cobra.Command) (*config.LocalConfigInfo, error) {
...
 		afs := FlagValueIfSet(command, ApplicationFlagName)
 		// Find the first child of the command. As some groups are allowed even with non existent config
 		fcc := getFirstChildOfCommand(command)
+		fmt.Println("fcc is", fcc.Name(), "command name is", command.Name())
```
Comment (kadel) [inline pkg/odo/genericclioptions/context.go]: I suspect that this is debugging output left accidentally

File: pkg/odo/genericclioptions/context.go (hunk 4 above)
Comment (surajnarwade) [reply pkg/odo/genericclioptions/context.go]: yeah :see_no_evil:

File: pkg/odo/genericclioptions/context.go (hunk 5 above)
Comment (surajnarwade) [reply pkg/odo/genericclioptions/context.go]: yeah :see_no_evil:

**Comment by surajnarwade** [comment]
@kadel , test case which I added don't use context flag anyway, so I think it should satisfy the case

**Comment by kadel** [comment]
> @kadel , test case which I added don't use context flag anyway, so I think it should satisfy the case

nope. It doesn't test the situation when you run the command from the directory without a component.
This needs to be tested. Basically, there needs to be a test that will fail without your last commit.

**Comment by surajnarwade** [comment]
@kadel ack 
I will check with @amitkrout and add this test

**Comment by kadel** [comment]
> @kadel ack
> I will check with @amitkrout and add this test

just add one test that calls `odo component list --path` from a directory that doesn't have a component (no .odo directory), doesn't have to be much more complicated than this

**Comment by surajnarwade** [comment]
@kadel , can you please review again :smile_cat:

File: tests/integration/json_test.go [hunk 6]
```diff
@@ -81,9 +81,11 @@ var _ = Describe("odojsonoutput", func() {
 			Expect(desiredSrorageList).Should(MatchJSON(actualSrorageList))
 
 			// odo list -o json --path .
-			pwd := helper.CmdShouldPass("pwd")
+			pwd := helper.Getwd()
 			desired := fmt.Sprintf(`{"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"Component","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"nodejs","creationTimestamp":null},"spec":{"type":"nodejs","source":"https://github.com/openshift/nodejs-ex"},"status":{"context":"%s","state":"Pushed"}}]}`, strings.TrimSpace(pwd))
-			actual := helper.CmdShouldPass("odo", "list", "-o", "json", "--path", ".")
+			helper.Chdir("/tmp")
```
Comment (kadel) [inline tests/integration/json_test.go]: this is not going to work on windows

Comment (kadel) [reply tests/integration/json_test.go]: you can use `helper.CreateNewContext()` instead of this. Don't forget to delete the dir

Comment (surajnarwade) [reply tests/integration/json_test.go]: ah okay

Comment (surajnarwade) [reply tests/integration/json_test.go]: done

File: tests/integration/json_test.go [hunk 7]
```diff
@@ -83,9 +83,11 @@ var _ = Describe("odojsonoutput", func() {
 			// odo list -o json --path .
 			pwd := helper.Getwd()
 			desired := fmt.Sprintf(`{"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"Component","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"nodejs","creationTimestamp":null},"spec":{"type":"nodejs","source":"https://github.com/openshift/nodejs-ex"},"status":{"context":"%s","state":"Pushed"}}]}`, strings.TrimSpace(pwd))
-			helper.Chdir("/tmp")
+			tmpDir := helper.CreateNewContext()
+			helper.Chdir(tmpDir)
 			actual := helper.CmdShouldPass("odo", "list", "-o", "json", "--path", pwd)
 			helper.Chdir(pwd)
+			helper.DeleteDir(tmpDir)
```
Comment (kadel) [inline tests/integration/json_test.go]: if the `CmdShouldPass` fails the `tmpDir` is not going to be deleted :-(

Comment (kadel) [reply tests/integration/json_test.go]: call `defer helper.DeleteDir(tmpDir)` right after `tmpDir := helper.CreateNewContext()`

File: pkg/odo/cli/component/list.go [hunk 8]
```diff
@@ -4,18 +4,19 @@ import (
...
-	"github.com/spf13/cobra"
-
 	"github.com/openshift/odo/pkg/component"
 	"github.com/openshift/odo/pkg/log"
 	appCmd "github.com/openshift/odo/pkg/odo/cli/application"
 	projectCmd "github.com/openshift/odo/pkg/odo/cli/project"
 	"github.com/openshift/odo/pkg/odo/genericclioptions"
 	odoutil "github.com/openshift/odo/pkg/odo/util"
+	"github.com/openshift/odo/pkg/odo/util/completion"
+	"github.com/pkg/errors"
```
Comment (kadel) [inline pkg/odo/cli/component/list.go]: why was `"github.com/pkg/errors"` and `"github.com/spf13/cobra"` moved here?

Comment (surajnarwade) [reply pkg/odo/cli/component/list.go]: just IDE things :disappointed:

File: tests/integration/json_test.go (hunk 7 above)
Comment (surajnarwade) [reply tests/integration/json_test.go]: updated

File: pkg/odo/cli/component/list.go (hunk 8 above)
Comment (surajnarwade) [reply pkg/odo/cli/component/list.go]: updated

**Comment by amitkrout** [comment]
Integration test looks good to me
/lgtm

**Pull request merged**